test/test_dynamic_transform.cpp

```diff
@@ -1001,6 +1007,135 @@ TEST_F(NVFuserTest, DynamicPadShmoo_CUDA) {
+TEST_F(NVFuserTest, DynamicRewriteEmptyTensors1_CUDA) {
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto x = makeSymbolicTensor({-1, 1});
+  auto y = makeSymbolicTensor({1, -1});
+  fusion.addInput(x);
+  fusion.addInput(y);
+
+  auto z = mul(x, y);
+  fusion.addOutput(z);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  at::Tensor at_x = at::randn({3, 1}, options);
+  at::Tensor at_y = at::randn({1, 0}, options);
+  std::vector<c10::IValue> aten_inputs = {at_x, at_y};
+
+  FusionExecutorCache fusion_executor_cache(std::move(fusion_ptr));
+  auto outputs = fusion_executor_cache.runFusionWithInputs(aten_inputs);
+
+  testValidate(
+      fusion_executor_cache.fusion(),
+      outputs,
+      aten_inputs,
+      {at_x * at_y},
+      __LINE__,
+      __FILE__);
+
+  auto runtime = fusion_executor_cache.getMostRecentKernelRuntime();
+  auto runtime_outputs = runtime->fusionSegments()->outputs();
+  EXPECT_EQ(runtime_outputs.size(), 1);
+
+  auto tv_output = runtime_outputs.at(0);
+  auto def = tv_output->definition();
+  EXPECT_NE(def, nullptr);
+  EXPECT_EQ(def->isA<FullOp>(), true);
+}
+
+// Similar to above, but no broadcasts. Instead perform a reduction over
+// non-empty axes first.
+TEST_F(NVFuserTest, DynamicRewriteEmptyTensors2_CUDA) {
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+
+  auto x = makeSymbolicTensor({-1, -1});
+  fusion.addInput(x);
+
+  auto y = sum(x, {1});
+  fusion.addOutput(y);
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+  at::Tensor at_x = at::randn({0, 3}, options);
+  std::vector<c10::IValue> aten_inputs = {at_x};
+
+  FusionExecutorCache fusion_executor_cache(std::move(fusion_ptr));
+  auto outputs = fusion_executor_cache.runFusionWithInputs(aten_inputs);
+
+  testValidate(
+      fusion_executor_cache.fusion(),
+      outputs,
+      aten_inputs,
+      {at_x.sum(1)},
+      __LINE__,
+      __FILE__);
+
+  auto runtime = fusion_executor_cache.getMostRecentKernelRuntime();
+  auto runtime_fusion = runtime->fusionSegments();
+  auto runtime_outputs = runtime_fusion->outputs();
+  EXPECT_EQ(runtime_outputs.size(), 1);
+
+  auto tv_output = runtime_outputs.at(0)->as<TensorView>();
+  auto def = tv_output->definition();
+  EXPECT_NE(def, nullptr);
+  EXPECT_EQ(def->isA<FullOp>(), true);
+
+  // Fusion output should have hardcoded zero extent after concretization
+  auto output_extent = tv_output->axis(0)->extent();
+  EXPECT_TRUE(output_extent->isConstInt());
+  EXPECT_EQ(output_extent->getInt(), 0);
+
+  // Fusion input should have hardcoded zero extent after concretization
+  auto input_extent =
+      runtime_fusion->inputs().at(0)->as<TensorView>()->axis(0)->extent();
+  EXPECT_TRUE(input_extent->isConstInt());
+  EXPECT_EQ(input_extent->getInt(), 0);
```

Comment:
This currently fails since I have not yet figured out how to replace inputs to a `Fusion`. Outputs are easy to replace using `replaceOutput`. We probably just need to add a simple `replaceInput` method.